utils/core/src/serde/byte_reader.rs

```diff
@@ -0,0 +1,166 @@
+
+    /// Returns true if there are more bytes left to be read from `self`.
+    fn has_more_bytes(&self) -> bool;
+
+    // PROVIDED METHODS
+    // --------------------------------------------------------------------------------------------
+
+    /// Returns a boolean value read from `self` consuming 1 byte from the reader.
+    ///
+    /// # Errors
+    /// Returns a [DeserializationError] if a u16 value could not be read from `self`.
+    fn read_bool(&mut self) -> Result<bool, DeserializationError> {
+        let byte = self.read_u8()?;
+        match byte {
+            0 => Ok(false),
+            1 => Ok(true),
+            _ => Err(DeserializationError::InvalidValue(format!(
+                "{byte} is not a boolean value"
+            ))),
+        }
+    }
+
+    /// Returns a u16 value read from `self` in little-endian byte order.
+    ///
+    /// # Errors
+    /// Returns a [DeserializationError] if a u16 value could not be read from `self`.
+    fn read_u16(&mut self) -> Result<u16, DeserializationError> {
+        let bytes = self.read_array::<2>()?;
+        Ok(u16::from_le_bytes(bytes))
+    }
+
+    /// Returns a u32 value read from `self` in little-endian byte order.
+    ///
+    /// # Errors
+    /// Returns a [DeserializationError] if a u32 value could not be read from `self`.
+    fn read_u32(&mut self) -> Result<u32, DeserializationError> {
+        let bytes = self.read_array::<4>()?;
+        Ok(u32::from_le_bytes(bytes))
+    }
+
+    /// Returns a u64 value read from `self` in little-endian byte order.
+    ///
+    /// # Errors
+    /// Returns a [DeserializationError] if a u64 value could not be read from `self`.
+    fn read_u64(&mut self) -> Result<u64, DeserializationError> {
+        let bytes = self.read_array::<8>()?;
+        Ok(u64::from_le_bytes(bytes))
+    }
+
+    /// Returns a u128 value read from `self` in little-endian byte order.
+    ///
+    /// # Errors
+    /// Returns a [DeserializationError] if a u128 value could not be read from `self`.
+    fn read_u128(&mut self) -> Result<u128, DeserializationError> {
+        let bytes = self.read_array::<16>()?;
+        Ok(u128::from_le_bytes(bytes))
+    }
+
+    /// Reads a deserializable value from `self`.
+    ///
+    /// # Panics
+    /// Panics if the value could not be read from `self`.
+    fn read<D: Deserializable>(&mut self) -> Result<D, DeserializationError>
+    where
+        Self: Sized,
+    {
+        D::read_from(self)
+    }
+}
+
+// SLICE READER
+// ================================================================================================
+
+/// Implements [ByteReader] trait for a slice of bytes.
+pub struct SliceReader<'a> {
+    source: &'a [u8],
+    pos: usize,
+}
+
+impl<'a> SliceReader<'a> {
+    /// Creates a new slice reader from the specified slice.
+    pub fn new(source: &'a [u8]) -> Self {
+        SliceReader { source, pos: 0 }
+    }
```

Comment:
One alternative is to hold only the slice that we effectively need. That way, we could get rid of `pos`, and [slice::split_at](https://doc.rust-lang.org/std/primitive.slice.html#method.split_at) whenever we are reading new content. Example:

Suppose we need to read 8 bytes
```rust
self.check_eor(8)?;
let (read, remainder) = self.source.split_at(8);
self.source = remainder;
(... do something with read ...)
```

Or, to peek:

```rust
(n < self.source.len()).then(|| &self.source[..n])
```

Also, as a constructor, it might be better to have more flexibility as we could set as argument a vec, slice, string, etc

```rust
pub fn new<B>(source: B) -> Self where B: AsRef<[u8]>
```

Reply:
This is a cool way to do it! However, I'm not sure how the compiler will handle this (i.e., if it will actually result in less efficient code) - so, without further investigation, I'll leave it as is for now. We can always improve this later as it will be non-breaking change.

Regarding a more general constructor: I tried to make it work but ran into lifetime bound issues - so, left it for later as well.